### ISSUE TITLE
CodeDOM does not work without the mono runtime in bundled executables

### ISSUE THREAD

**Issue by Wertzui123** [mono/mono#20767]
## Steps to Reproduce

1. Paste this code into a file and compile it:
```
using System.CodeDom.Compiler;
using System.IO;
using Microsoft.CSharp;

namespace Example
{
    internal static class Example
    {

        public static void Main()
        {
            Compile();
        }

        private static void Compile()
        {
            var provider = new CSharpCodeProvider();
            File.WriteAllText(Path.Combine(Directory.GetCurrentDirectory(), "HelloWorld.cs"), @"
			using System;

namespace App
{
    internal static class Program
    {
        
        public static void Hello()
        {
            Console.WriteLine(""Hello World!"");
        }
        
    }
}
");
            provider.CompileAssemblyFromFile(new CompilerParameters
            {
                GenerateExecutable = false,
                GenerateInMemory = false,
                OutputAssembly = "HelloWorld.dll",
            }, Path.Combine(Directory.GetCurrentDirectory(), "HelloWorld.cs"));
        }

    }
}
```
2. Create a native executable with `mkbundle Example.exe -o Example --static --deps --simple --cross default`
3. Run the "Example" file **on a blank debian without mono**

### Current Behavior

An exception is thrown: System.SystemException: Error running mcs: Cannot find the specified file

### Expected Behavior

The assembly should compile successfully.

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**:

Mono JIT compiler version 4.6.2 (Debian 4.6.2.7+dfsg-1)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        LLVM:          supported, not enabled.
        GC:            sgen

## Stacktrace

```
Unhandled Exception:
System.SystemException: Error running mcs: Cannot find the specified file
  at Mono.CSharp.CSharpCodeCompiler.CompileFromFileBatch (System.CodeDom.Compiler.CompilerParameters options, System.String[] fileNames) [0x00176] in <bd46d4d4f7964dfa9beea098499ab597>:0
  at Mono.CSharp.CSharpCodeCompiler.CompileAssemblyFromFileBatch (System.CodeDom.Compiler.CompilerParameters options, System.String[] fileNames) [0x00011] in <bd46d4d4f7964dfa9beea098499ab597>:0
  at System.CodeDom.Compiler.CodeDomProvider.CompileAssemblyFromFile (System.CodeDom.Compiler.CompilerParameters options, System.String[] fileNames) [0x00014] in <bd46d4d4f7964dfa9beea098499ab597>:0
  at Example.Example.Compile () [0x00057] in <29f0fc9349ea410693d471d09f0ed8e3>:0
  at Example.Example.Main () [0x00000] in <29f0fc9349ea410693d471d09f0ed8e3>:0
```

**Comment by lambdageek** [comment]
That is a limitation of how CodeDom works on mono:

the source for the CSharpCodeProvider is in https://github.com/mono/corefx/blob/master/src/System.CodeDom/src/Microsoft/CSharp/CSharpCodeProvider.cs  it delegates most work to `CSharpCodeGenerator` which is in https://github.com/mono/mono/blob/master/mcs/class/System/Microsoft.CSharp/CSharpCodeGenerator.cs  and as you can see it tries to run `mcs` from somewhere on your path

**Comment by Wertzui123** [comment]
Is there any way to remove this dependency? I try to make a 100% independent program.

**Comment by Wertzui123** [comment]
So, bundling mcs.exe into the executable works, but the error still occurs. Is there any way to use the mcs.exe inside my bundled executable to compile the script?

**Comment by lambdageek** [comment]
> So, bundling mcs.exe into the executable works, but the error still occurs. Is there any way to use the mcs.exe inside my bundled executable to compile the script?

I don't think that will work.  I don't think there's a way to make `Process.Start` run an .exe from inside the bundle. which is what it would take to make CodeDom work.

**Comment by Wertzui123** [comment]
I wonder how mono manages to find and load the assemblies within a bundle. Maybe I can replicate this haivuor myself.

**Comment by lambdageek** [comment]
> I wonder how mono manages to find and load the assemblies within a bundle. Maybe I can replicate this behavior myself.

There's a difference between starting a new process and how mono loads an assembly into the running process.  The way mono opens a bundled assembly is by using `mmap` to map the assembly (which is embedded within the bundled executable) into the address space of the running process.

**Comment by Wertzui123** [comment]
Wouldn't it be possible to use this to extract the mcs.exe?

**Comment by Picoseconds** [comment]
Cant you do that to a forked process and run mcs?

**Comment by Picoseconds** [comment]
I think you should use a resource and then monkey-patch Process.Start to use that one which you temporaty extraact
